File: java-streams/examples/java/src/main/java/streams/ChainAndIntermediateTerminal.java
package streams;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ChainAndIntermediateTerminal {

    Stream<Employee> findEmployees(List<Person> persons) {

        return persons.stream()
                .filter(person -> person.isEmployed())
                .map(person -> new Employee(person.getFirstName() + " " + person.getLastName()));
    }

    Stream<String> employeeNames(Stream<Employee> employees) {
        return employees.map(Employee::getName); // intermediate
    }

    List<String> nameList(List<Person> persons) {
        return employeeNames(findEmployees(persons))
                .collect(Collectors.toList()); // terminal
    }
}

File: java-streams/examples/kotlin/src/main/kotlin/streams/ChainAndIntermediateTerminal.kt
package streams

class ChainAndIntermediateTerminal {

    fun findEmployees(persons: Sequence<Person>) =
            persons.asSequence()
                    .filter(Person::isEmployed) // intermediate
                    .map { person -> Employee("${person.firstName} ${person.lastName}") } // intermediate

    fun employeeNames(employees: Sequence<Employee>) =
            employees.map { it.name } // intermediate

    fun nameList(persons: List<Person>) =
            employeeNames(findEmployees(persons.asSequence())).toList() // terminal

    fun nameList2(persons: List<Person>) =
            persons.asSequence()
                    .let { findEmployees(it) }
                    .let { employeeNames(it) }
                    .toList() // terminal

}
